Mathlib/NumberTheory/NumberField/FinitePlaces.lean

```diff
@@ -0,0 +1,242 @@
+      (zero_le ((toNNReal (norm_ne_zero v)) (v.valuation y)))
+    have h_mono := StrictMono.monotone (toNNReal_strictMono (one_lt_norm v))
+    rw [← Monotone.map_max h_mono] --max goes inside withZeroMultIntToNNReal
+    exact h_mono (Valuation.map_add v.valuation x y)
+
+theorem vadic_abv_def {x : K} : vadic_abv v x = toNNReal (norm_ne_zero v) (v.valuation x) := rfl
+
+end absoluteValue
+
+section FinitePlace
+variable {K : Type*} [Field K] [NumberField K] (v : HeightOneSpectrum (𝓞 K))
+
+/-- The embedding of a NumberField inside its completion with respect to `v`. -/
+def embedding : K →+* adicCompletion K v :=
+    @UniformSpace.Completion.coeRingHom K _ v.adicValued.toUniformSpace _ _
+
+noncomputable instance instRankOneValuedAdicCompletion :
+    Valuation.RankOne (valuedAdicCompletion K v).v where
+  hom := {
+    toFun := toNNReal (norm_ne_zero v)
+    map_zero' := rfl
+    map_one' := rfl
+    map_mul' := MonoidWithZeroHom.map_mul (toNNReal (norm_ne_zero v))
+  }
+  strictMono' := toNNReal_strictMono (one_lt_norm v)
+  nontrivial' := by
+    rcases Submodule.exists_mem_ne_zero_of_ne_bot v.ne_bot with ⟨x, hx1, hx2⟩
+    use (x : K)
+    rw [valuedAdicCompletion_eq_valuation' v (x : K)]
+    constructor
+    · simpa only [ne_eq, map_eq_zero, NoZeroSMulDivisors.algebraMap_eq_zero_iff]
+    · apply ne_of_lt
+      rw [valuation_eq_intValuationDef, intValuation_lt_one_iff_dvd]
+      exact dvd_span_singleton.mpr hx1
+
+/-- The `v`-adic completion of `K` is a normed field. -/
+noncomputable instance instNormedFieldValuedAdicCompletion : NormedField (adicCompletion K v) :=
+    Valued.toNormedField (adicCompletion K v) (WithZero (Multiplicative ℤ))
+
+/-- A finite place of a number field `K` is a place associated to an embedding into a completion
+with respect to a maximal ideal. -/
+def FinitePlace (K : Type*) [Field K] [NumberField K] := {w : AbsoluteValue K ℝ //
+    ∃ v : IsDedekindDomain.HeightOneSpectrum (𝓞 K), place (embedding v) = w}
+
+/-- Return the finite place defined by a maximal ideal `v`. -/
+noncomputable def FinitePlace.mk (v : HeightOneSpectrum (𝓞 K)) : FinitePlace K :=
+    ⟨place (embedding v), ⟨v, rfl⟩⟩
+
+/-- The norm of the image after the embedding associated to `v` is equal to the `v`-adic absolute
+value. -/
+theorem FinitePlace.norm_def (x : K) : ‖embedding v x‖ = vadic_abv v x := by
+  simp only [NormedField.toNorm, instNormedFieldValuedAdicCompletion, Valued.toNormedField,
+    instFieldAdicCompletion, Valued.norm, Valuation.RankOne.hom, MonoidWithZeroHom.coe_mk,
+    ZeroHom.coe_mk, embedding, UniformSpace.Completion.coeRingHom, RingHom.coe_mk,
+    MonoidHom.coe_mk, OneHom.coe_mk, Valued.valuedCompletion_apply, NNReal.coe_inj]
+  rfl
+
+open FinitePlace
+
+/-- The `v`-adic norm of an integer is at most 1. -/
+theorem norm_le_one (x : 𝓞 K) : ‖embedding v x‖ ≤ 1 := by
+  rw [norm_def, vadic_abv_def, NNReal.coe_le_one, toNNReal_le_one_iff (one_lt_norm v)]
```

Comment:
The opening gambits of this and the next two proofs are quite similar. It may be a good idea to state a lemma for this and then use it in these proofs.